traefik/templates/service.yaml

```diff
@@ -1,25 +1,33 @@
 {{- if .Values.service.enabled -}}
 
-{{ $tcpPorts := dict }}
-{{ $udpPorts := dict }}
-{{ $exposedPorts := false }}
-{{- range $name, $config := .Values.ports }}
-  {{- if or $config.http3 (eq (toString $config.protocol) "UDP") }}
-    {{ $_ := set $udpPorts $name $config }}
-  {{- end }}
-  {{- if eq (toString (default "TCP" $config.protocol)) "TCP" }}
-    {{ $_ := set $tcpPorts $name $config }}
-  {{- end }}
-  {{- if (eq $config.expose true) }}
-    {{ $exposedPorts = true }}
-  {{- end }}
-{{- end }}
+{{- $tcpPorts := dict -}}
+{{- $udpPorts := dict -}}
+{{- $exposedPorts := false -}}
+{{- range $name, $config := .Values.ports -}}
+  {{- if $config.http3 -}}
+    {{/* Due to https://github.com/kubernetes/kubernetes/issues/58477 it is not possible
+      to define the HTTP/3 UDP port as a (named) containerPort. TCP and UDP containerPort
+      will have the same value for 'port' which will break the chart upgrade. Therefore
+      we manually refer to port number for HTTP/3. */}}
+    {{- $http3Name := (toString $config.port) }}
+    {{- $_ := set $udpPorts $http3Name $config.http3 -}}
+  {{- end -}}
+  {{- if eq (toString $config.protocol) "UDP" -}}
+    {{ $_ := set $udpPorts $name $config -}}
+  {{- end -}}
```

Comment:
I think this mostly comes down to how you want to structure the port values in `values.yaml`.
The `http3` config is now nested within `websecure`, because they are so tightly related.

This causes it not to be considered when listing container-ports on the pod, which is neccessary, as listing two ports with the same port number is currently not possible (in a reliable way) in Kubernetes due to merging mechanics. Therefore it is not a named pod-port like the others (e.g. `web` or `websecure`), so it requires a reference by port number instead of port name.

For all other ports, including other UDP-ports, there is no problem in referring to them by name. You would still need to be careful not to define multiple named ports that share the same port number, but I think that's practically a different problem. Disabling port-by-name

Reply:
My use-case does not involve any HTTP/3, but DNS. Would this be supported with the proposed change?

Reply:
> My use-case does not involve any HTTP/3, but DNS. Would this be supported with the proposed change?

This change does not actually change or add any functionality; all it does is merge all TCP/UDP ports into a single service by default, where previously it would create 2 seperate services. 

The old split-services behavior can be kept with the `service.single: false` value; see also the upgrade notes I added to the README in this PR.

Reply:
> all it does is merge all TCP/UDP ports into a single service by default, where previously it would create 2 seperate services.

That's perfect for DNS, as it listens on TCP _and_ UDP 53 on the same IP address aka LoadBalancer.

Reply:
It would be perfect but in K8S it is not possible to listen TCP and UDP on the same port AFAIK. So your use case would be impossible to implement in k8s

Reply:
Yeah, for http3 this PR explicitly does not use 443 and instead uses an alternative port (8443), as using the same port number for both UDP and TCP [does not fully work in Kubernetes](https://github.com/kubernetes/kubernetes/issues/39188). Creating from scratch will *probably* work, but any edited config might not merge correctly due to this.